File: 0046-permutations/0046-permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();

        backtrack(new ArrayList<>(), nums, ans);

        return ans;
    }

    private void backtrack (List<Integer> current, int[] nums, List<List<Integer>> ans) {
        if (current.size() == nums.length) ans.add(new ArrayList<>(current));
        else {
            for (int i = 0; i < nums.length; i++) {
                if (current.contains(nums[i])) continue;
                current.add(nums[i]);
                backtrack(current, nums, ans);
                current.remove(current.size() - 1);
            }
        }
    }
}

File: 0046-permutations/0046-permutations.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
    let ans = [];

    backtrack(0, nums, ans);

    return ans;
};

function backtrack (start, nums, ans) {
    if (start === nums.length) {
        ans.push([...nums]);
    } else {
        for (let i = start; i < nums.length; i++) {
            swap(nums, start, i);
            backtrack(start + 1, nums, ans);
            swap(nums, start, i);
        }
    }
}

function swap (nums, a, b) {
    let temp = nums[a];
    nums[a] = nums[b];
    nums[b] = temp;
}